// nestjs-typeorm-paginate/dist/create-pagination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createPaginationObject = void 0;
const pagination_1 = require("./pagination");
function createPaginationObject({ items, totalItems, currentPage, limit, route, metaTransformer, routingLabels, }) {
    const totalPages = Math.ceil(totalItems / limit);
    const hasFirstPage = route;
    const hasPreviousPage = route && currentPage > 1;
    const hasNextPage = route && currentPage < totalPages;
    const hasLastPage = route && totalPages > 0;
    const symbol = route && new RegExp(/\?/).test(route) ? '&' : '?';
    const limitLabel = routingLabels && routingLabels.limitLabel
        ? routingLabels.limitLabel
        : 'limit';
    const pageLabel = routingLabels && routingLabels.pageLabel ? routingLabels.pageLabel : 'page';
    const routes = {
        first: hasFirstPage ? `${route}${symbol}${limitLabel}=${limit}` : '',
        previous: hasPreviousPage
            ? `${route}${symbol}${pageLabel}=${currentPage - 1}&${limitLabel}=${limit}`
            : '',
        next: hasNextPage
            ? `${route}${symbol}${pageLabel}=${currentPage + 1}&${limitLabel}=${limit}`
            : '',
        last: hasLastPage
            ? `${route}${symbol}${pageLabel}=${totalPages}&${limitLabel}=${limit}`
            : '',
    };
    const meta = {
        totalItems: totalItems,
        itemCount: items.length,
        itemsPerPage: limit,
        totalPages: totalPages,
        currentPage: currentPage,
    };
    if (metaTransformer)
        return new pagination_1.Pagination(items, metaTransformer(meta), route && routes);
    // @ts-ignore
    return new pagination_1.Pagination(items, meta, route && routes);
}
exports.createPaginationObject = createPaginationObject;
