// nestjs-typeorm-paginate/dist/paginate.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.paginateRawAndEntities = exports.paginateRaw = exports.paginate = void 0;
const typeorm_1 = require("typeorm");
const interfaces_1 = require("./interfaces");
const create_pagination_1 = require("./create-pagination");
const DEFAULT_LIMIT = 10;
const DEFAULT_PAGE = 1;
function paginate(repositoryOrQueryBuilder, options, searchOptions) {
    return __awaiter(this, void 0, void 0, function* () {
        return repositoryOrQueryBuilder instanceof typeorm_1.Repository
            ? paginateRepository(repositoryOrQueryBuilder, options, searchOptions)
            : paginateQueryBuilder(repositoryOrQueryBuilder, options);
    });
}
exports.paginate = paginate;
function paginateRaw(queryBuilder, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const [page, limit, route, paginationType] = resolveOptions(options);
        const totalQueryBuilder = queryBuilder.clone();
        const [items, total] = yield Promise.all([
            (paginationType === interfaces_1.PaginationTypeEnum.LIMIT_AND_OFFSET
                ? queryBuilder.limit(limit).offset((page - 1) * limit)
                : queryBuilder.take(limit).skip((page - 1) * limit)).getRawMany(),
            totalQueryBuilder.getCount(),
        ]);
        return create_pagination_1.createPaginationObject({
            items,
            totalItems: total,
            currentPage: page,
            limit,
            route,
            metaTransformer: options.metaTransformer,
            routingLabels: options.routingLabels,
        });
    });
}
exports.paginateRaw = paginateRaw;
function paginateRawAndEntities(queryBuilder, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const [page, limit, route, paginationType] = resolveOptions(options);
        const totalQueryBuilder = queryBuilder.clone();
        const [itemObject, total] = yield Promise.all([
            (paginationType === interfaces_1.PaginationTypeEnum.LIMIT_AND_OFFSET
                ? queryBuilder.limit(limit).offset((page - 1) * limit)
                : queryBuilder.take(limit).skip((page - 1) * limit)).getRawAndEntities(),
            totalQueryBuilder.getCount(),
        ]);
        return [
            create_pagination_1.createPaginationObject({
                items: itemObject.entities,
                totalItems: total,
                currentPage: page,
                limit,
                route,
                metaTransformer: options.metaTransformer,
                routingLabels: options.routingLabels,
            }),
            itemObject.raw,
        ];
    });
}
exports.paginateRawAndEntities = paginateRawAndEntities;
function resolveOptions(options) {
    const page = resolveNumericOption(options, 'page', DEFAULT_PAGE);
    const limit = resolveNumericOption(options, 'limit', DEFAULT_LIMIT);
    const route = options.route;
    const paginationType = options.paginationType || interfaces_1.PaginationTypeEnum.LIMIT_AND_OFFSET;
    return [page, limit, route, paginationType];
}
function resolveNumericOption(options, key, defaultValue) {
    const value = options[key];
    const resolvedValue = Number(value);
    if (Number.isInteger(resolvedValue) && resolvedValue >= 0)
        return resolvedValue;
    console.warn(`Query parameter "${key}" with value "${value}" was resolved as "${resolvedValue}", please validate your query input! Falling back to default "${defaultValue}".`);
    return defaultValue;
}
function paginateRepository(repository, options, searchOptions) {
    return __awaiter(this, void 0, void 0, function* () {
        const [page, limit, route] = resolveOptions(options);
        if (page < 1) {
            return create_pagination_1.createPaginationObject({
                items: [],
                totalItems: 0,
                currentPage: page,
                limit,
                route,
                metaTransformer: options.metaTransformer,
                routingLabels: options.routingLabels,
            });
        }
        const [items, total] = yield repository.findAndCount(Object.assign({ skip: limit * (page - 1), take: limit }, searchOptions));
        return create_pagination_1.createPaginationObject({
            items,
            totalItems: total,
            currentPage: page,
            limit,
            route,
            metaTransformer: options.metaTransformer,
            routingLabels: options.routingLabels,
        });
    });
}
function paginateQueryBuilder(queryBuilder, options) {
    return __awaiter(this, void 0, void 0, function* () {
        const [page, limit, route, paginationType] = resolveOptions(options);
        const [items, total] = yield (paginationType ===
            interfaces_1.PaginationTypeEnum.LIMIT_AND_OFFSET
            ? queryBuilder.limit(limit).offset((page - 1) * limit)
            : queryBuilder.take(limit).skip((page - 1) * limit)).getManyAndCount();
        return create_pagination_1.createPaginationObject({
            items,
            totalItems: total,
            currentPage: page,
            limit,
            route,
            metaTransformer: options.metaTransformer,
            routingLabels: options.routingLabels,
        });
    });
}
